solidity/ecdsa/contracts/libraries/DKG.sol

```diff
@@ -0,0 +1,508 @@
+        bytes groupPubKey;
+        // Array of misbehaved members indices (disqualified or inactive).
+        // Indices must be in range [1, groupSize], unique, and sorted in ascending
+        // order.
+        uint8[] misbehavedMembersIndices;
+        // Concatenation of signatures from members supporting the result.
+        // The message to be signed by each member is keccak256 hash of the
+        // calculated group public key, misbehaved members indices and DKG
+        // start block. The calculated hash should be prefixed with prefixed with
+        // `\x19Ethereum signed message:\n` before signing, so the message to
+        // sign is:
+        // `\x19Ethereum signed message:\n${keccak256(
+        //    groupPubKey, misbehavedMembersIndices, dkgStartBlock
+        // )}`
+        bytes signatures;
+        // Indices of members corresponding to each signature. Indices must be
+        // be in range [1, groupSize], unique, and sorted in ascending order.
+        uint256[] signingMembersIndices;
+        // Identifiers of candidate group members as outputted by the group
+        // selection protocol.
+        uint32[] members;
+        // Keccak256 hash of group members identifiers that actively took part
+        // in DKG (excluding IA/DQ members).
+        bytes32 membersHash;
+    }
+
+    /// @notice States for phases of group creation. The states doesn't include
+    ///         timeouts which should be tracked and notified individually.
+    enum State {
+        // Group creation is not in progress. It is a state set after group creation
+        // completion either by timeout or by a result approval.
+        IDLE,
+        // Group creation is awaiting the seed and sortition pool is locked.
+        AWAITING_SEED,
+        // Off-chain DKG protocol execution is in progress. A result is being calculated
+        // by the clients in this state. It's not yet possible to submit the result.
+        KEY_GENERATION,
+        // After off-chain DKG protocol execution the contract awaits result submission.
+        // This is a state to which group creation returns in case of a result
+        // challenge notification.
+        AWAITING_RESULT,
+        // DKG result was submitted and awaits an approval or a challenge. If a result
+        // gets challenge the state returns to `AWAITING_RESULT`. If a result gets
+        // approval the state changes to `IDLE`.
+        CHALLENGE
+    }
+
+    /// @dev Size of a group in the threshold relay.
+    uint256 public constant groupSize = 100;
+
+    /// @notice Time in blocks after which DKG result is complete and ready to be
+    //          published by clients.
+    // TODO: In the ECDSA implementation a fixed length of offchain DKG time
+    // won't be used. Rework the solution to await a DKG result submission
+    // once a key generation started.
+    uint256 public constant offchainDkgTime = 5 * (1 + 5) + 2 * (1 + 10) + 20;
```

Comment:
I see the TODO comment, but for future us, I think it's cheaper to put a calculated const value without any math. If we want to add math for explanation, we can do that in comments.